internal redirects: support passing headers from response to request

pugmajere opened the pull request:
Commit Message:
This adds a new (repeated) field in the internal redirect policy, "response_headers_to_preserve".  When set, the headers named there will be copied from the response that triggers an internal redirect into the request that follows.

This allows some limited information passing through the internal redirect system.

The current system is faithful to the idea that internal redirects are purely a latency optimization, and should behave similarly to if the redirect had been passed to the downstream user-agent. This does violate that idea.

Other proxies, such as Nginx, have a much more flexible way of handling internal redirects that allows a fair bit of information passing like this. This should allow implementations to adopt Envoy that are using this kind of information passing, with reduced needs to rearchitect.

Risk Level: Low
Testing: Various tests added. Manual test done as well.
Docs Changes: Docs updated to reflect the additional features.
Release Notes: Release notes updated.
Fixes: #30441
Fixes: #16777

API Considerations: Requires configuration, so default is no behavior change. Otherwise, I believe everything else is covered or not relevant.

pugmajere commented:
I'm not particularly excited about the field name, so I'm happy to take suggestions for a better name if desired.

Also, I feel like I don't understand the lifetimes for the header maps very well, so I defaulted to using SetCopy(). I'm happy to change those if that helps at all.

pugmajere commented:
At least one of the failed checks looks to be a flake in the test infrastructure, btw.

mattklein123 commented:
/lgtm api

alyssawilk commented on docs/root/intro/arch_overview/http/http_connection_management.rst:
```diff
@@ -252,6 +256,11 @@ and then shipped upstream with all of the normal Envoy request sanitization taki
   avoid duplicating undesired header values.
 
+.. warning::
+  Note that no filters will see the response that triggers the internal redirect. If there is a need
```
No _downstream_ filters will.  You actually can see and modify headers with upstream filters

pugmajere replied:
Right.  I'll update with that.

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -1837,15 +1838,61 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
+
+  for (const auto& to_preserve :
+       route_entry_->internalRedirectPolicy().responseHeadersToPreserve()) {
+    auto header = Http::LowerCaseString(to_preserve);
```
why store them as string instead of having a vector of LowerCaseString in the first place?

pugmajere replied:
Right, that makes sense.  No point in doing that conversion on every request.

alyssawilk commented on api/envoy/config/route/v3/route_components.proto:
```diff
@@ -2391,6 +2392,14 @@ message InternalRedirectPolicy {
   bool allow_cross_scheme_redirect = 4;
+
+  // Specifies a list of headers, by name, to copy from the internal redirect into the subsequent
+  // request. If a header is specified here but not present in the redirect, it will be cleared in
```
Hm, I can defer to the  API shepherds, but I'd find it unexpected to say "preserve foo" and if foo was not sent, have it erase an existing foo header.

@mattklein123 saw you signed off but can I get your thoughts or @htuch here?

mattklein123 replied:
Yeah that's a good point. Perhaps call is response_headers_to_copy? I think we use that name elsewhere and might better cover what happens.

pugmajere replied:
That works for me.

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -1837,15 +1838,61 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
+       route_entry_->internalRedirectPolicy().responseHeadersToPreserve()) {
+    auto header = Http::LowerCaseString(to_preserve);
+    auto result = upstream_headers.get(header);
+    auto dresult = downstream_headers.get(header);
```
I think we can fully qualify downstream_

pugmajere replied:
Done.

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -1837,15 +1838,61 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
+    auto header = Http::LowerCaseString(to_preserve);
+    auto result = upstream_headers.get(header);
+    auto dresult = downstream_headers.get(header);
+    if (result.empty()) {
```
I'm going to pause here and wait for API review, so I only review final functionality :-)

pugmajere replied:
I think I've caught up on all the comments now.

alyssawilk commented on api/envoy/config/route/v3/route_components.proto:
```diff
@@ -2391,6 +2392,14 @@ message InternalRedirectPolicy {
   bool allow_cross_scheme_redirect = 4;
+
+  // Specifies a list of headers, by name, to copy from the internal redirect into the subsequent
+  // request. If a header is specified here but not present in the redirect, it will be cleared in
```
Usually for headers we explicitly disallow special HTTP/2 headers (:host :scheme etc).  Again question for API shepherds if we want to be consistent here?

mattklein123 replied:
I think it probably depends on the use case. If we want to block that in code yeah we should document here.

pugmajere replied:
There's not much point in adding those into this list, since they're already copied.... but there's also no sense in letting users do silly and confusing things.

Is there a standard pattern for disallowing these that I should follow?

alyssawilk replied:
for config validation we use HeaderUtility::isModifiableHeader and reject those if configured.  not seeing useful comments to crib off of.

pugmajere replied:
added a check using isModifiableHeader in config_impl.cc; I'm reasonably sure that's the right place.

alyssawilk reviewed (commented):
Matt, adding you back for 2 more API comments, then you can clear again :-)

alyssawilk commented:
/wait on CI and comments

pugmajere commented:
I've been fighting a bit with the pre-push checks because they don't work on M2 Macs, which is my primary development environment for this work.  I think things will pass now, at least.

alyssawilk commented on source/common/router/config_impl.h:
```diff
@@ -591,6 +592,7 @@ class InternalRedirectPolicyImpl : public InternalRedirectPolicy {
   const absl::flat_hash_set<Http::Code> redirect_response_codes_;
   std::vector<std::pair<InternalRedirectPredicateFactory*, ProtobufTypes::MessagePtr>>
       predicate_factories_;
+  std::vector<Http::LowerCaseString> response_headers_to_copy_;
```
comment please

pugmajere replied:
Added.

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -1840,15 +1841,59 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
-          downstream_headers.setScheme(scheme_is_http ? Http::Headers::get().SchemeValues.Http
-                                                      : Http::Headers::get().SchemeValues.Https);
+
+  auto saved_headers = Http::RequestHeaderMapImpl::create();
```
Honestly at this point if we need to restore original headers I think it'd be easier to just clone and restore, than try to unwind.  WDYT?

alyssawilk replied:
If we stick with this let's make sure is thoroughly tested

pugmajere replied:
Yeah, that is less code, and easier to follow, I think. Done.

alyssawilk reviewed (commented):
LG modulo a few remaining nits
/wait

pugmajere commented:
I have rebased to try to deal with the repeated conflict resolution I was doing.
In effect, I've added another 2 commits to the stack, so you can see the changes easily, still, I hope.

(Also, my commit messages after the first have been written assuming they will be squashed away.  If that's not the case, well, I can clean things up a bit, heh.)

mattklein123 reviewed (approved):
API LGTM

pugmajere commented on source/common/router/router.cc:
```diff
@@ -1837,17 +1838,50 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
+
+        saved_headers->iterate(
+            [&downstream_headers](const Http::HeaderEntry& header) -> Http::HeaderMap::Iterate {
+              downstream_headers.addCopy(Http::LowerCaseString(header.key().getStringView()),
```
It won't, but I think that matches the behavior we get with chained internal redirects and requestHeadersToAdd anyway, isn't it?

HeaderMapImpl is marked NonCopyable, so std::move doesn't really work here without equivalent levels of API changes.

Maybe replaceFrom() that calls clear() and copyFrom() internally would be a reasonable addition?

I'm not really sure what to do here either.

alyssawilk replied:
what if we made this a copyFrom as well, and I later on fix copyFrom to clear first?

pugmajere replied:
That works for me.
Done!

pugmajere commented on source/common/router/router.cc:
```diff
@@ -1837,17 +1838,50 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
-  const std::string original_path(downstream_headers.getPathValue());
   const bool scheme_is_set = (downstream_headers.Scheme() != nullptr);
+
+  auto saved_headers = Http::RequestHeaderMapImpl::create();
```
Done.

pugmajere commented on source/common/router/config_impl.cc:
```diff
@@ -368,6 +368,12 @@ InternalRedirectPolicyImpl::InternalRedirectPolicyImpl(
     Envoy::Config::Utility::translateOpaqueConfig(predicate.typed_config(), validator, *config);
     predicate_factories_.emplace_back(&factory, std::move(config));
   }
+  for (const auto& header : policy_config.response_headers_to_copy()) {
```
Done.

pugmajere commented on source/common/router/config_impl.h:
```diff
@@ -578,6 +578,7 @@ class InternalRedirectPolicyImpl : public InternalRedirectPolicy {
   }
 
   std::vector<InternalRedirectPredicateSharedPtr> predicates() const override;
+  std::vector<Http::LowerCaseString> responseHeadersToCopy() const override;
```
That sounded like a great idea (though I doubt it makes much of a difference in practice), but wow, I had to touch a bunch of tests because references don't play as nicely with MOCK_METHOD.

It's done, at least.

pugmajere commented on source/common/router/router.cc:
```diff
@@ -1837,17 +1838,45 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
+
+  for (const auto& header : route_entry_->internalRedirectPolicy().responseHeadersToCopy()) {
+    auto result = upstream_headers.get(header);
+    auto downstream_result = downstream_headers.get(header);
```
Done.

RyanTheOptimist commented:
LGTM, though it looks like this needs a main merge.

alyssawilk reviewed (commented):
LGTM modulo one nit and the main merge

RyanTheOptimist commented:
/wait

pugmajere commented:
Fixed the merge conflicts in the changelog since I think this is hopefully done now.

pugmajere commented:
I don't think the latest changes caused any of these test failures, so if someone can retrigger the tests, that would be great.

(The tests failed after I did the conflict-merge for the changelog, which ... seems unlikely to cause the failures seen, lol.)

pugmajere commented:
/retry-azp

(Not sure if that will work, but worth a shot)

alyssawilk commented:
/retest :-)
